reframe/core/meta.py

```diff
@@ -93,6 +95,48 @@ def __getitem__(self, key):
                         # raise the exception from the base __getitem__.
                         raise err from None
 
+    class WrappedFunction:
+        '''Descriptor to wrap a free function as a bound-method.
+
+        The free function object is wrapped by the constructor. Instances
+        of this class should be inserted into the namespace of the target class
+        with the desired name for the bound-method. Since this class is a
+        descriptor, the `__get__` method will return the right bound-method
+        when accessed from a class instance.
+        '''
+
+        __slots__ = ('fn')
+
+        def __init__(self, fn, name=None):
+            @functools.wraps(fn)
+            def fn_(*args, **kwargs):
+                return fn(*args, **kwargs)
+
+            self.fn = fn_
+            if name:
+                self.fn.__name__ = name
+
+        def __get__(self, obj, objtype=None):
+            if objtype is None:
+                objtype = type(obj)
+
+            self.fn.__qualname__ = '.'.join(
+                [objtype.__qualname__, self.fn.__name__]
+            )
+            if obj is None:
+                return self.fn
+
+            return types.MethodType(self.fn, obj)
+
+        def __getattr__(self, name):
+            return getattr(self.fn, name)
+
+        def __setattr__(self, name, value):
```

Comment:
Why are you using `__setattr__` here and not `__set__`?

Reply:
I think this should remain as a non-data descriptor. The method `__setattr__` is to attach attributes into `self.fn`; whereas `__set__` would be used to change `self.fn` itself, which is something that Python does not allow. For example:
```python
class Foo:
  def fn(self):
    print('fn')

  def __init__(self):
    self.fn = 4 # This inserts 'fn' directly in self.__dict__['fn'], because the method fn is a non-data descriptor.

f = Foo()
print(f.fn) # Prints 4 and fn is stored under self.__dict__['fn'].
```
I just noticed that this is missing the `__call__` method to be a "proper" Python function object.